pkg/controller/stacks/host/config.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package host
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// EnvTenantKubeconfig is the environment variable pointing to kubeconfig file of custom resource Kubernetes API
+	// (a.k.a Tenant Kubernetes). This environment variable is the main switch enabling Stack Manager hosted mode.
+	EnvTenantKubeconfig = "TENANT_KUBECONFIG"
+
+	envControllerNamespace         = "CONTROLLER_NAMESPACE"
+	envTenantKubernetesServiceHost = "TENANT_KUBERNETES_SERVICE_HOST"
+	envTenantKubernetesServicePort = "TENANT_KUBERNETES_SERVICE_PORT"
+
+	errMissingEnvVar = "host aware mode activated but %s env var is not set"
+)
+
+// HostedConfig is the configuration for Host Aware Mode where different Kubernetes API's are used for pod
+// scheduling and custom resources.
+type HostedConfig struct {
+	// HostControllerNamespace is the namespace on Host Cluster where install and controller jobs/deployments will be
+	// deployed.
+	HostControllerNamespace string
+	// TenantAPIServiceHost is Kubernetes Apiserver Host for custom resources (a.k.a Tenant Kubernetes)
+	TenantAPIServiceHost string
+	// TenantAPIServicePort is Kubernetes Apiserver Port for custom resources (a.k.a Tenant Kubernetes)
+	TenantAPIServicePort string
+}
+
+// NewHostedConfig returns a new HostAwareConfig based on the available environment variables.
+func NewHostedConfig() (*HostedConfig, error) {
+	tenantKubeconfig := os.Getenv(EnvTenantKubeconfig)
+	if tenantKubeconfig == "" {
+		// Hosted Mode is off, regular installation.
+		return nil, nil
+	}
+	// Hosted Mode is on, we need all the following environment variables to be set.
+	ns := os.Getenv(envControllerNamespace)
```

Comment:
Yes, it make sense and sounds better. 
The reason that I did like this is not to flow this parameter set down to this function from main.go. For example, I will need to change signature of [`setupWithManagerFunc`](https://github.com/crossplaneio/crossplane/pull/1139/files#diff-66525c8a65d872f540928ed763ef9d54R92) to accept extra config. 
Do you think it worths doing so? Or do you see a cleaner way?

Reply:
I agree that the plumbing could be a bit awkward, but typically I think it is best to have all config loaded in one place (in or close to `main.go`) and plumbed down to where it's used. This makes it easier for code readers to figure out where the config comes from, and can make it easier to override config in tests. So I do think it's worth doing. 

I would not shy away from solutions that involve refactoring our current `main.go` implementation (to whatever extent is necessary, including getting rid of the pluggable `setupWithManager` pattern). I think we arrived at the current implementation somewhat organically, rather than because anyone really loves it.

One pattern I like to use in this situation is to add [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) to my controller setup function, then build those options from Kingpin flags in `main.go`.  You can see examples of this in [crossplane-tools](https://github.com/crossplaneio/crossplane-tools/blob/9033bd/cmd/angryjet/main.go) and [draino](https://github.com/planetlabs/draino/blob/a877e7f0852/cmd/draino/draino.go).

Reply:
Thanks a lot for the links.
Now, we are getting the config options via flags as you suggested.